app/services/recommendation_engine.rb

```diff
@@ -1,23 +1,13 @@
 class RecommendationEngine
-  def initialize(favorite_movies)
-    @favorite_movies = favorite_movies
+  def initialize(params)
+    @user_id = params[:user_id]
   end
 
   def recommendations
-    movie_titles = get_movie_names(@favorite_movies)
-    genres = Movie.where(title: movie_titles).pluck(:genre)
+    user = User.find(@user_id)
+    rented = user.rented
+    genres = Movie.where(id: user.favorites).pluck(:genre)
     common_genres = genres.group_by{ |e| e }.sort_by{ |k, v| -v.length }.map(&:first).take(3)
-    Movie.where(genre: common_genres).order(rating: :desc).limit(10)
-  end
-
-  private
-
-  def get_movie_names(movies)
-    names = []
-    @favorite_movies.each do |movie|
-      names << movie.title
-    end
-
-    return names
+    Movie.by_genre(common_genres).where.not(id: rented).order(rating: :desc).limit(10)
```

Comment:
These improvements on recommendation engine not only makes it simpler and more readable, but also faster  on ActiveRecord and less Allocations. Since allocating in Ruby is slow, we can make our programs faster by removing allocations.